en/lessons/basics/collections.md

```diff
@@ -3,209 +3,774 @@ version: 1.3.1
+this means you are going to see them everywhere in Haskell. They are generally
+the first container a Haskell programmer reaches for.  Because of slow append
+and relatively slow indexed access (O(n) where n is the index) they are
+generally used in situations where you know you are going to have to iterate
+over the entire data set, or you want to preserve the order of elements.
+
+A great example of an abstract data structure for which a linked list is a great
+concrete implementation are stacks. This is because pushing and popping are
+O(1).
+
+A bad use case would be for a queue, where either enqueue or dequeue would be
+O(n) depending on which side of the linked list you decide to insert into.
+
+A common real world example of where lists are used is database queries. A
+database query could return no results `[]`, or some `[entity..]`, and there is
+potentially an ordering to these results. The database library doesn't really
+care about indexed access, and so it leaves that consideration to the caller of
+the function.
+
 ### List Concatenation
 
-List concatenation uses the `++/2` operator:
+List concatenation uses the `++` operator:
 
-```elixir
-iex> [1, 2] ++ [3, 4, 1]
+```haskell
+ghci> [1, 2] ++ [3, 4, 1]
 [1, 2, 3, 4, 1]
 ```
 
-A side note about the name (`++/2`) format used above:
-In Elixir (and Erlang, upon which Elixir is built), a function or operator name has two components: the name you give it (here `++`) and its _arity_.
-Arity is a core part of speaking about Elixir (and Erlang) code.
-It is the number of arguments a given function takes (two, in this case).
-Arity and the given name are combined with a slash. We'll talk more about this later; this knowledge will help you understand the notation for now.
+A side note about another operator that you might come across `<>`. This
+operator is sometimes referred to as `mappend`, it represents generic
+concatenation. In other words any type that supports `<>` can be "appended". In
+the case of lists `<>` is exactly `++` under the hood.
 
-### List Subtraction
+### Head / Tail
+
+When using lists, it is common to work with a list's head and tail. The head is
+the list's first element, while the tail is a list containing the remaining
+elements.
 
-Support for subtraction is provided via the `--/2` operator; it's safe to subtract a missing value:
+Haskell provides two helpful functions, `head` and `tail`, for working with
+these parts:
 
-```elixir
-iex> ["foo", :bar, 42] -- [42, "bar"]
-["foo", :bar]
+```haskell
+ghci> head ["Orange", "Banana", "Apple"]
+"Orange"
+ghci> tail ["Orange", "Banana", "Apple"]
+["Banana","Apple"]
 ```
 
-Be mindful of duplicate values.
-For every element on the right, the first occurrence of it gets removed from the left:
+Unfortunately these functions reveal an ugly part of the language's base
+library, some of the functions are partial. This means that they do not cover
+the full domain of possible inputs.
```

Comment:
Maybe combine the two? "may raise an exception, even when given an argument with an appropriate type, because they do not cover the full domain of possible inputs" this way we include both the cause and the effect